Home_Work_2.py

```diff
@@ -0,0 +1,178 @@
+from abc import ABC, abstractmethod
+from functools import wraps
+
+
+class BookModel(BaseModel):
+    name: str
+    author: str
+    issue_date: str
+
+
+class AbstractClass(ABC):
+    @abstractmethod
+    def __init__(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def __str__(self):
+        raise NotImplementedError
+
+
+class Book(AbstractClass):
+    name = None
+    __author = None
+    issue_date = None
+    book_list = list()
+    book_dict = dict()
+
+    def __init__(self, book: BookModel):
+        self.name = book.name
+        self.__author = book.author
+        self.issue_date = book.issue_date
+        self.book_list.append(self.__repr__())
+        self.book_dict.setdefault(self.__author, self.name)
+
+    def __str__(self):
+        return (f'Book "{self.name.title()}" was written by {self.__author.title()} '
+                f'and issued {self.issue_date}.')
+
+    def __repr__(self):
+        return f'{self.name.title()}'
+
+
+class Magazine(Book):
+
+    def __init__(self, book: BookModel, style):
+        super().__init__(book)
+        self.style = style
+
+    def __str__(self):
+        return (f'Magazine "{self.name.title()}" was written by {self.__author.title()},'
+                f' magazine style is {self.style} and issued {self.issue_date}.')
+
+
+# decorator for adding book
+def add_book_decorator(func):
+    @wraps(func)
+    def inner(self, *args, **kwargs):
+        print(f'\nNew book "{args[1]}" was succesfully added to Library')
+        return func(self, *args, **kwargs)
+
+    return inner
+
+
+# decorator for deleting book
+def del_book_decorator(func):
+    @wraps(func)
+    def inner(self, *args, **kwargs):
+        if args[0] in self.book_dict.keys():
+            print(f'\nBook "{self.book_dict[args[0]]}" was succesfully deleted from Library')
+            return func(self, *args, **kwargs)
+
+    return inner
+
+
+class Library():
+
+    def __init__(self):
+        self.book_list = Book.book_list
+        self.book_dict = Book.book_dict
+
+    def __iter__(self):
+        self.__index = -1
+        return self
+
+    def __next__(self):
+        self.__index += 1
+        if self.__index < len(self.book_list):
+            return self.book_list[self.__index]
+        else:
+            raise StopIteration
+
+    def __getitem__(self, item):
+        return self.book_dict.get(item, 'Wrong data, try again!!')
+
+    # ТУТ не понял, ведь генератор вощврашает каждый следующий елемент, а как вернуть по имени автора?
+    # Потому сделал выше ^ как смог через обычный словарь и __gettattr__
+    # def __getitem__(self, item):
+    #     res = iter((item[1] for item in self.book_dict.items()))
+    #     return next(res)
+
+    @del_book_decorator
+    def __delitem__(self, key):
+        del self.book_dict[key]
+        print(self.book_dict)
+
+    @add_book_decorator
+    def __setitem__(self, author, book):
+        self.book_dict[author] = book
+        print(self.book_dict)
```

Comment:
I like you idea, but it will be more easy to just append a book to the list of books